### PR TITLE
Expose JSDoc tags through the language service

### LINKED ISSUE #11956: changed behavior of symbol.getDocumentationComment()
I am interested in expanding the use of JsDoc comments to annotate interface fields with validation criteria. An attempt to use symbol.getDocumentationComment() seems to have different behavior
between 2.0 and 2.1 branches.

Sample doc comment (attached to an interface here, but I don't think that should matter)
 ```
export interface Foo {
 /**
  * a description string
  *
  * @customTag foo
  * @customElse bar
  */
 field: string;
};

```

The intent is to walk the tree of nodes and invoke getDocumentationComment on the symbol for 'field'
```
(<ts.Symbol>field).getDocumentationComment()

```

**Expected behavior:**
```
// version typescript@2.0.6

[ { text: 'a description string', kind: 'text' },
  { text: '\n', kind: 'lineBreak' },
  { text: '', kind: 'text' },
  { text: '\n', kind: 'lineBreak' },
  { text: '@customTag foo', kind: 'text' },
  { text: '\n', kind: 'lineBreak' },
  { text: '@customElse bar', kind: 'text' } ]
```

**Actual behavior:** 
```
// version typescript@2.1.0-dev.20161031
[ { text: 'a description string', kind: 'text' }]
```

If this is intended behavior, please provide a way to get raw documentation so client applications and extend any default usage.

### PR DESCRIPTION
**Pull request by minestarks**
Expose JSDoc tags through the language service

The main motivation behind this PR is to fix UWP API documentation behavior in Dev15, which relies on a custom JSDoc tag (`@xmldoc`), and was broken between 2.0 and 2.1.

The behavior of getDocumentationComment() was changed by https://github.com/Microsoft/TypeScript/pull/10671  to not include JSDoc tags anymore. I added them back as a new property.

Note the "tags" property I added to these calls will only expose the unknown tags (i.e. things that are not specially handled like `@returns,` `@param` etc) ... I couldn't think of a justification to expose the known tags too (well, also I wouldn't have been able to do so without making more significant changes to the parser which I want to avoid at the moment).

Fixes #11956 

You might be interested @sandersn  @vladima

### CODE REVIEW

File: src/services/jsDoc.ts [hunk 1]
```diff
@@ -68,6 +68,33 @@ namespace ts.JsDoc {
...
+    export function getJsDocTagsFromDeclarations(declarations: Declaration[]) {
+        // Only collect doc comments from duplicate declarations once.
+        const tags: JSDocTagInfo[] = [];
+        forEachUnique(declarations, declaration => {
+            const jsDocs = getJSDocs(declaration);
+            if (!jsDocs) {
+                return;
+            }
+            const jsDocTags: JSDocTag[] = [];
+            jsDocs.forEach(doc => {
```
Comment (sandersn) [inline src/services/jsDoc.ts]: Why not just `for (const doc in jsDocs)`?

File: src/services/types.ts [hunk 2]
```diff
@@ -492,12 +494,18 @@ namespace ts {
         kind: string;
     }
 
+    export interface JSDocTagInfo {
```
Comment (sandersn) [inline src/services/types.ts]: I'm not sure about the naming scheme in services, but I'd prefer a terser name, like JSDocTag.

File: src/compiler/parser.ts [hunk 3]
```diff
@@ -6468,7 +6468,7 @@ namespace ts {
 
                 function parseTagComments(indent: number) {
                     const comments: string[] = [];
-                    let state = JSDocState.SawAsterisk;
```
Comment (sandersn) [inline src/compiler/parser.ts]: Is this just a drive-by bug fix? I didn't see any real difference in parsing functionality.

File: src/services/jsDoc.ts [hunk 4]
```diff
@@ -68,6 +68,33 @@ namespace ts.JsDoc {
...
+                const tagsForDoc = (doc as JSDoc).tags;
```
Comment (vladima) [inline src/services/jsDoc.ts]: is type assertion necessary?

File: src/services/jsDoc.ts [hunk 5]
```diff
@@ -68,6 +68,33 @@ namespace ts.JsDoc {
...
+                if (tagsForDoc) {
+                    jsDocTags.push(...tagsForDoc.filter(tag => tag.kind === SyntaxKind.JSDocTag));
+                }
+            });
+            if (!jsDocTags) {
```
Comment (vladima) [inline src/services/jsDoc.ts]: always false since array is pre-allocated

File: src/services/jsDoc.ts [hunk 6]
```diff
@@ -68,6 +68,33 @@ namespace ts.JsDoc {
...
+                return;
+            }
+            for (const tag of jsDocTags) {
```
Comment (vladima) [inline src/services/jsDoc.ts]: silly question: do we need intermediate `jsDocTags`? Can we just put data into `tags` directly?

File: src/services/jsDoc.ts [hunk 7]
```diff
@@ -68,6 +68,33 @@ namespace ts.JsDoc {
...
+                if (tag) {
```
Comment (vladima) [inline src/services/jsDoc.ts]: can it ever be `null` or `undefined`? I think in this case line `83` should fail

File: src/compiler/parser.ts (hunk 3 above)
Comment (minestarks) [inline src/compiler/parser.ts]: Ah yes I should clarify. This isn't irrelevant. It fixes a bug I discovered while running the new quickinfo test I added. Consider this comment:

```ts
/**
  * @mytag1 
  * here all the comments are on a new line
  * @mytag2
  */
```

When the comment parser starts out at "SawAsterisk" state I get this tag comment for  mytag1: `"* here all the comments are on a new line"`

With my change it's fixed: `"here all the comments are on a new line"`

This is because we skip all the whitespace (including newlines) before we begin parsing so we start right at the *. This would have been handled in the AsteriskToken case below, but only if the last state had been BeginningOfLine. 

Going over the whole logic I believe it just makes more sense this way. Starting at BeginningOfLine is more logical since we didn't actually see an asterisk so why start at SawAsterisk...

Comment (sandersn) [inline src/compiler/parser.ts]: Makes sense. There was a similar bug in the start state for the main jsdoc parsing, so it makes sense that you ran into it here too.

File: src/services/jsDoc.ts (hunk 4 above)
Comment (minestarks) [inline src/services/jsDoc.ts]: I guess not. I got confused by the 2.0 language service giving me an implicit any warning...

File: src/services/services.ts [hunk 8]
```diff
@@ -404,6 +416,10 @@ namespace ts {
         // symbol has no doc comment, then the empty string will be returned.
         documentationComment: SymbolDisplayPart[];
 
+        // Undefined is used to indicate the value has not been computed. If, after computing, the
+        // symbol has no doc comment, then the empty array will be returned.
+        jsDocTags: JSDocTagInfo[];
```
Comment (mihailik) [inline src/services/services.ts]: Empty array is expensive here.

Most of `SignatureObject` instances will have no JSDoc comments. Allocating empty arrays will add bunch of fluff to GC. Would be best to avoid that.

File: src/services/services.ts [hunk 9]
```diff
@@ -427,6 +443,14 @@ namespace ts {
 
             return this.documentationComment;
         }
+
+        getJsDocTags(): JSDocTagInfo[] {
+            if (this.jsDocTags === undefined) {
+                this.jsDocTags = this.declaration ? JsDoc.getJsDocTagsFromDeclarations([this.declaration]) : [];
+            }
```
Comment (mihailik) [inline src/services/services.ts]: How about using `undefined` for storing un-computed state and `null` for computed and empty?

File: src/services/jsDoc.ts [hunk 10]
```diff
@@ -68,6 +68,28 @@ namespace ts.JsDoc {
         return documentationComment;
     }
 
+    export function getJsDocTagsFromDeclarations(declarations: Declaration[]) {
+        // Only collect doc comments from duplicate declarations once.
+        const tags: JSDocTagInfo[] = [];
```
Comment (mihailik) [inline src/services/jsDoc.ts]: Here too the array may be lazy-allocated when first tag is encountered -- meaning no allocation for the default case.

File: src/services/services.ts [hunk 11]
```diff
@@ -316,6 +320,14 @@ namespace ts {
 
             return this.documentationComment;
         }
+
+        getJsDocTags(): JSDocTagInfo[] {
+            if (this.tags === undefined) {
+                this.tags = JsDoc.getJsDocTagsFromDeclarations(this.declarations);
+            }
```
Comment (mihailik) [inline src/services/services.ts]: Here too the check would need to change to avoid empty array allocation for the default no-JSDoc case.
(to mark the difference between un-computed value and empty value in `this.tags` -- `undefined` versus `null` can be used)

File: src/services/services.ts [hunk 12]
```diff
@@ -1315,7 +1339,8 @@ namespace ts {
                                 kindModifiers: ScriptElementKindModifier.none,
                                 textSpan: createTextSpan(node.getStart(), node.getWidth()),
                                 displayParts: typeToDisplayParts(typeChecker, type, getContainerNode(node)),
-                                documentation: type.symbol ? type.symbol.getDocumentationComment() : undefined
+                                documentation: type.symbol ? type.symbol.getDocumentationComment() : undefined,
+                                tags: type.symbol ? type.symbol.getJsDocTags() : undefined
```
Comment (mihailik) [inline src/services/services.ts]: Note that the API may return `undefined` here, so returning `null` or `undefined` instead of an empty array shouldn't break expectations.

File: tests/baselines/reference/quickInfoDisplayPartsArrowFunctionExpression.baseline [hunk 13]
```diff
@@ -73,7 +73,8 @@
           "kind": "keyword"
         }
       ],
-      "documentation": []
+      "documentation": [],
+      "tags": []
```
Comment (mihailik) [inline tests/baselines/reference/quickInfoDisplayPartsArrowFunctionExpression.baseline]: And now look at those billions of empty arrays peppering the ***.baseline down from here.
